### PR TITLE
Add initial LGPL VAX disassembler with tests for MOVL, NOP, BRB, and …

### PR DESCRIPTION
**Pull request by ghost**
**Your checklist for this pull request**  
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository.  
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style).  
- [x] I've documented every `RZ_API` function and struct this PR changes.  
- [x] I've added tests that prove my changes are effective.  
- [x] I've updated the [Rizin book](https://github.com/rizinorg/book) with the relevant information (if needed).  

**Detailed description**  
Draft PR for my GSoC 2025 project **"Rewriting GPL-only code"** for **VAX** in Rizin.  
Replaces GPL-licensed VAX disassembler with a new LGPL implementation.  

- Added `librz/arch/vax/` with support for **7 instructions**:  
  `HALT, MOVL, NOP, BRB, ADDL, PUSHL, CALLS`.  
- Includes tests in `test/db/arch/vax/` using a **NetBSD binary** (`ls.vax`) and manually crafted binaries.  
- Seeking early feedback per mentor **(xvilka)** guidance.  

**Test plan**  

**Build Rizin:**  
```bash
meson setup build && cd build && ninja
```

**Run VAX tests:**  
```bash
cd test && rz-test db/arch/vax
```
- **Expected:** All tests (`vax_halt`, `vax_movl`, etc.) pass.  

**Manually test:**  
```bash
rizin -A test/bins/vax/ls.vax
ie
pd 1 @ 0x120001000  # Should show: 0x120001000      04           halt
```

**Test movl.bin:**  
```bash
rizin -a vax test/bins/vax/movl.bin
pd 1  # Should show: 0x00000000      d0505050     movl r0, r0
```

**Edge case: Invalid opcode:**  
```bash
echo -ne '\xff' > test/bins/vax/invalid.bin
rizin -a vax test/bins/vax/invalid.bin
pd 1  # Should show: 0x00000000      ff           invalid
```

**Closing issues**  
N/A (draft PR for GSoC project, addressing **GPL removal for VAX**).

### CODE REVIEW

File: test/bin/vax/movl.bin [hunk 1]
```diff
@@ -0,0 +1 @@
+echo -ne '\xd0\x50\x50' > ../bin/vax/movl.bin
```
Comment (notxvilka) [inline test/bin/vax/movl.bin]: It's not done like that, see `test/db/asm/*` as an example.  E.g. in `test/db/asm/msp430` one line looks like:
```
d "bic #2, r10" 2ac3 0x0 (set r10 (& (~ (bv 16 0x2)) (var r10)))
```
`d` means disassembly, then go instruction mnemonic, then bytes, then offset. The last part is RzIL, which you can omit in this plugin.

**Review by notxvilka — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
i'll fix them asap

File: librz/arch/meson.build [hunk 2]
```diff
@@ -512,3 +530,6 @@ modules += { 'rz_arch': {
     ],
     'plugins': [arch_plugins]
 }}
+
+
+
```
Comment (Rot127) [inline librz/arch/meson.build]:
```suggestion
```

Remove unnecessary lines.

File: librz/arch/vax/vax.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+// Copyright (C) 2025 Rizin
+
+#ifndef RZ_VAX_H
+#define RZ_VAX_H
+
+typedef struct {
+    const char *name;  // e.g., "movl"
+    unsigned int opcode;  // e.g., 0xD0
+    const char *args;  // e.g., "rlwl"
+} vax_instruction;
```
Comment (Rot127) [inline librz/arch/vax/vax.h]: Please use CamelCase for type names.

File: librz/arch/vax/vax_disasm.c [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    if (!insn->name) {
+        rz_strbuf_set(sb, "invalid");
+        return opcode_len;
+    }
+
+    // Print the instruction name
+    rz_strbuf_setf(sb, "%s", insn->name);
+
+    // Process operands based on args
+    if (insn->args[0]) {
```
Comment (Rot127) [inline librz/arch/vax/vax_disasm.c]: Negate condition and return early. `if (!insn->args[0]) { return... }`

File: librz/arch/vax/vax_disasm.c [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        const ut8 *p = buf + opcode_len;
```
Comment (Rot127) [inline librz/arch/vax/vax_disasm.c]: Rename `p` to something more descriptive please.

File: librz/arch/vax/vax_disasm.c [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        int arg_idx = 0;
+        while (insn->args[arg_idx]) {
+            if (arg_idx > 0) rz_strbuf_append(sb, ", ");
+            char mode = p[0];
+            p++;
+            ut8 reg = mode & 0xF;
+
+            if (strcmp(insn->name, "brb") == 0) {
```
Comment (Rot127) [inline librz/arch/vax/vax_disasm.c]:
```suggestion
            if (RZ_STR_EQ(insn->name, "brb")) {
```

Comment (Rot127) [reply librz/arch/vax/vax_disasm.c]: But you should have some enum identifiers for this. Using strings as identifiers is inefficient.

File: librz/arch/vax/vax_disasm.c [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    // Print the instruction name
+    rz_strbuf_setf(sb, "%s", insn->name);
+
+    // Process operands based on args
+    if (insn->args[0]) {
+        const ut8 *p = buf + opcode_len;
+        int arg_idx = 0;
+        while (insn->args[arg_idx]) {
```
Comment (Rot127) [inline librz/arch/vax/vax_disasm.c]: `args` seem to encode the operands the instruction has? But it is not used below, can you elaborate why?

**Review by Rot127 — CHANGES_REQUESTED** [review]

File: test/bin/vax/brb.bin
Comment (wargio) [inline test/bin/vax/brb.bin:1]: These needs to be pushed into the bin repository

**Review by wargio — CHANGES_REQUESTED** [review]

**Review by ghost — COMMENTED** [review]
please review these changes , mentors

**Comment by wargio** [comment]
you have a very long way to go, before this is even reviewable:

1. you should not touch the VAX gnu plugin but instead making your own one.
2. do not create a bin for each instruction and instead read the documentation https://github.com/rizinorg/rizin/blob/dev/test/README.md#writing-assembly-tests
3. given the changes, i strongly suggest to start first getting used to the code, by fixing something easier, then you can start writing this.

File: librz/arch/p_gnu/asm/asm_vax_gnu.c [hunk 8]
```diff
@@ -21,62 +21,64 @@ static const ut8 *bytes = NULL;
...
-	int delta = (memaddr - Offset);
-	if (delta < 0) {
-		return -1; // disable backward reads
-	}
-	if (delta > length) {
-		return -1;
-	}
-	memcpy(myaddr, bytes + delta, RZ_MIN(length, bytes_size));
-	return 0;
+    int delta = (memaddr - Offset);
```
Comment (Rot127) [inline librz/arch/p_gnu/asm/asm_vax_gnu.c:24]: You need t run `clang-format` on those. The easiest way is `git-clang-format` once all the files are staged. Rizin uses currently `clang-format-16`

**Comment by notxvilka** [comment]
@thedvlprguy when you fix some comment, don't forget to press the "resolve" button.

**Comment by notxvilka** [comment]
Closing this because 1) no activity 2) no meaningful code changes since first it was opened.
